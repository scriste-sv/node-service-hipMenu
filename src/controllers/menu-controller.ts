import { Restaurant } from './../model/restaurant-model';

interface IMenu {
    _id: number,
    name: string
}

export class MenuController {

    //Menu
    public readMenu(id: number) {
        const query: Object = { "menu._id": id };

        return Restaurant.find(query)
            .then(entities => {
                return entities.map((list) => {
                    return (<any>list).menu.map((menu: IMenu) => {
                        if (menu._id == id) {
                            return menu;
                        }
                    }).pop();
                }).pop();
            }) 
            .catch(err => {
                return err;
            });
    };

    public readAllMenus() {
        return Restaurant.find()
            .then(entities => {
                const result = [].concat.apply([], entities.map((list) => {
                    return (<any>list).menu.map((menu: IMenu) => {
                        return menu;
                    });
                }));
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

    public createMenu(id: string, data: Object) {
        const query: Object = {_id: id};

        return Restaurant.updateOne(query, {$push : {menu: data}})
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

    public updateMenu(id: string, data: {name: string}) {
        const query: Object = { "menu": { "$elemMatch": { "_id": id }}};
        const condition: Object = { "$set": { "menu.$": {"_id": id , "name": data.name}}};


        return Restaurant.updateOne(query,condition)
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

    public deleteMenu(id: number) {
        const query: Object = { "menu": { "$elemMatch": { "_id": id }}};
        const condition: Object = { "$pull": { "menu": { "_id": id }}};

        return Restaurant.updateOne(query, condition)
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }
}