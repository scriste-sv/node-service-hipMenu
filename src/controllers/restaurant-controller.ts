import { Restaurant } from './../model/restaurant-model';

export class RestaurantController {

    //Restaurant
    public create(data: Object) {
        return Restaurant.create(data)
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

    public read(id: string) {
        const query: Object = { _id: id };

        return Restaurant.findOne(query)
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    } 

    public readAll() {
        return Restaurant.find()
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

    public update(id: string, data: Object) {
        const query: Object = { _id: id };

        return Restaurant.updateOne(query, data)
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

    public delete(id: string) {
        const query: Object = { _id: id };
        return Restaurant.findByIdAndDelete(query, {})
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

    public searchRestaurant(name: string) {
        const query: Object = { name: new RegExp(name, 'i') };

        return Restaurant.find(query)
            .then(result => {
                return result;
            }) 
            .catch(err => {
                return err;
            });
    }

}
