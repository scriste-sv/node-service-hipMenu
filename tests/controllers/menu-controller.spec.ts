import { MenuController } from '../../src/controllers/menu-controller';
import { Restaurant } from '../../src/model/restaurant-model';
import { objectTest, objectTestUpdated } from '../objectTest';

describe('Controller Restaurant', () => {
    let controller: MenuController;
    let objectMock: Object;

    beforeEach(() => {
        controller = new MenuController();
        objectMock = objectTest;
        Restaurant.create = jest.fn();
        Restaurant.findOne = jest.fn();
        Restaurant.find = jest.fn();
        Restaurant.updateOne = jest.fn();
        Restaurant.findByIdAndDelete = jest.fn();
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    // Menus

    it('returns a menu entity', async () => {
        (<any>objectMock).menu[2]._id = '123';
        (<jest.Mock>Restaurant.find).mockResolvedValue([objectMock]);

        const result = await controller.readMenu(123);

        expect(result).toEqual({"_id": "123", "name": "TEST"});
    });

    it('returns all menu entity', async () => {
        delete (<any>objectMock).menu[2]._id;
        (<jest.Mock>Restaurant.find).mockResolvedValue([objectMock]);

        const result = await controller.readAllMenus();

        expect(result).toEqual([{"name": "Desert"}, {"name": "Main dish"}, {"name": "TEST"}]);
    });

    it('creates and returns a menu entity', async () => {
        (<jest.Mock>Restaurant.updateOne).mockResolvedValue({name: 'TEST'});

        const result = await controller.createMenu('123', {name: 'TEST'});

        expect(result).toEqual({name: 'TEST'});
    });

    it('updates a menu entity', async () => {
        (<jest.Mock>Restaurant.updateOne).mockResolvedValue((<any>objectTestUpdated).menu[2]);

        const result = await controller.updateMenu('1234', (<any>objectMock).menu[2]);

        expect(result).toEqual((<any>objectTestUpdated).menu[2]);
    });
    
    it('deletes and returns a restaurant entity after an id search', async () => {
        (<jest.Mock>Restaurant.updateOne).mockResolvedValue(null);

        await controller.deleteMenu(1234);

        expect((<jest.Mock>Restaurant.updateOne).mock.calls[0]).toMatchSnapshot();
    });

});