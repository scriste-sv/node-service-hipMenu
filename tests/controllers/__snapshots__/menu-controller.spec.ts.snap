// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller Restaurant deletes and returns a restaurant entity after an id search 1`] = `
Array [
  Object {
    "menu": Object {
      "$elemMatch": Object {
        "_id": 1234,
      },
    },
  },
  Object {
    "$pull": Object {
      "menu": Object {
        "_id": 1234,
      },
    },
  },
]
`;
